packages/grafana-data/src/types/datasource.ts

```diff
@@ -25,6 +25,15 @@ export type DataSourceQueryType<DSType> = DSType extends DataSourceApi<infer TQu
 // Utility type to extract the options type TOptions from a class extending DataSourceApi<TQuery, TOptions>
 export type DataSourceOptionsType<DSType> = DSType extends DataSourceApi<any, infer TOptions> ? TOptions : never;
 
+export type QueryRelatedDataProviders = {
+  [name in string]: QueryRelatedDataProvider<any>;
+};
+
+export interface QueryRelatedDataProvider<TData> {
+  setRequest(request: DataQueryRequest): void;
+  getData(): Observable<TData>;
+}
```

Comment:
I would maybe add something like `getMetadata()/getQuery()` or something like that so you can get static info about what query will actually be executed by `getData`. That way in the future we can show user a little bit better what will happen, what will be run and also things like add the query to explore queries if that makes sense or do things like add hints to dashboard panel that would then create more panels for you based on queries/panels you already have.

Reply:
Good point. It'd might be nice to see in inspector page. Probably `getMetadata()` makes sense, with something like `queries` property (in theory there might be more than one).

Reply:
Hmm good point, so basically dataProvider does not represent single query but more of a piece of information that can be retrieved which can generate multiple queries is that good way to look at it?

Reply:
Yes. A DataProvider simply provides TData object. It can be anything and it may require a single query or multiple queries. Implementation details are up to the provider which is responsible for taking a request object and producing TData objects out of it.